docs/TypeScript.md

```diff
@@ -125,6 +125,32 @@ The type system heavily relies on generic properties to provide the most accurat
 
 🎉 Good work, now you can define interfaces for each route and have strictly typed request and reply instances. Other parts of the Fastify type system rely on generic properties. Make sure to reference the detailed type system documentation below to learn more about what is available.
 
+### Type checking responses based on the status code
+
+If you would like to be strict about the responses that are returned to the caller of a route. You can use the RepliesType when creating a route this key allows a type to be specified for any status code.  When `.sendReplyTyped` is called on the reply object, both the status code and the payload data is type scheduled to match.
+
+   ```typescript
+   server.get<{
+    RepliesTyped: {
```

Comment:
we already have a schema property, could you derive those from there?

Reply:
In my understanding the schema property for a path is the union of all possible result types for that path. So we'd first have to figure out how to break up the schema information so it can apply per status code.

Secondarily, the schema property is actual "data" and a JSON schema (actual "data" meaning it is not removed by `tsc` compile time. The JSON schema is not interpretable as a type definition in Typescript's perspective. Typescript doesn't interpret literal "data" as type information in my experience without a lot more complication.

I've seen some work in this direction but its mostly around using validation libraries like [Yup](https://github.com/jquense/yup) and [Zod](https://github.com/colinhacks/zod).

Here is the documentation from Yup showing how a schema expressed in Yup can built a set of assertions that match it.

https://github.com/jquense/yup/blob/master/docs/typescript.md

Zod offers the same type of support for the schemas that are defined using it.

There is also a module that converts Zod typed to JSON schemas.  [zod-to-json-schema](https://github.com/StefanTerdell/zod-to-json-schema)

If you're willing to consider changing the current JSON schema definitions with one of these libraries that offer Typescript definitions from the schema spec it may be possible to unify these definitions.